src/use_debounced_callback/index.jsx

```diff
@@ -1,11 +1,18 @@
 import debounce from 'lodash/debounce';
-import { useMemo } from 'react';
-
-const useDebouncedCallback = (func, delay) => (
-  useMemo(
-    () => debounce(func, delay),
-    [func, delay],
-  )
-);
+import { useEffect, useState } from 'react';
+
+const useDebouncedCallback = (func, delay) => {
+  const [instance, setInstance] = useState(undefined);
+
+  useEffect(() => {
+    const debounced = debounce(func, delay);
+
+    setInstance(debounced);
+
+    return () => { debounced.flush(); };
```

Comment:
https://lodash.com/docs/4.17.15#debounce

`cancel()` vs `flush()`

I think we should use `flush` here. I wouldn't want to loose previous calls only because either `func` or `delay` changes. The one thing I want to prevent is calling old `func` after we already created a new instance of debounce.

Reply:
I'm not so sure, we could get re-render called with props swapped and component re-used (ex: list items) in which case you would immediately fire a callback e.g. resize on an unmounting node. This is not so helpful, canceling is safer and faster.

Also what ablay pointed out - not having cancel() called, this hook could abstract it away but I'm not sure if the old one is broken? useMemo ... was returning a debounced function exposing .cancel() so it was up to you to cancel the functtion returned by the hook. In any case if we're going for convenience why not like this:

```javascript
import throttleOrDebounce from 'lodash';
import { useEffect, useRef } from 'react';

const useFancyWhatever = (fn, delay) => {
  const callback = useRef(throttleOrDebounce(fn, delay))

  useEffect(() => {
    callback.current = throttleOrDebounce(fn, delay);
    // Re-rendering, can discard previous calls
    return () => { callback.current.cancel(); };
  }, [fn, delay]);

  return callback.current;
};

export default useFancyWhatever;
```

This will execute server side, but i'm not sure if it's such a problem.

Reply:
About the example first (will reply to the rest later): I don't think this will work pretty good because the effect will never trigger a new render because you return a ref. Then you cannot use the returned callback as an effect dep - which is pretty unexpected from an hook

Reply:
Sorry, I don't get your re-render example - you still need to call the returned function one time for it to do anything

Reply:
I wouldn't say the old one is 'broken'. In my opinion, @atomiomi is right, the hook abstracting debounce should manage its lifecycle. Components using `useDebouncedCallback` shouldn't need to think about cases like this - we pretty much have everything we need here to cover this case.

But still a big issue: We trying to cover a case here, for which we have no real test case - not inside an app nor in our test suite. Maybe we should park this PR and continue with it when it's causing real-world issues?

Reply:
> I don't think this will work pretty good because the effect will never trigger a new render because you return a ref. Then you cannot use the returned callback as an effect dep - which is pretty unexpected from an hook

I see your point, useEffect will just update local ref container silently, and since we expose ref.current to the outer scope, we will always have handler that is *1 version behind* in the DOM. You can still use the returned callback as you like. React doesn't care where the dependency value comes from to fire useEffect

What we want to do is simply cancel all pending calls on the previous handler in case of changes. That means:
1. immediately expose current handler wrapped in debounce
2. cache current handler
3. cancel pending calls on changes.

See this revised example:

```javascript
const useDebouncedCallback = (fn, delay = 1000) => {
  const wrapped = debounce(fn, delay);
  const cache = useRef(wrapped)

  useEffect(() => {
    cache.current = wrapped;
    return () => { cache.current.cancel(); };
  }, [wrapped, fn, delay]); // `wrapped` here is redundant

  return wrapped;
};
```

> But still a big issue: We trying to cover a case here, for which we have no real test case - not inside an app nor in our test suite. Maybe we should park this PR and continue with it when it's causing real-world issues?

You're right we don't have a test case because we don't test components. That doesn't mean it isn't a real world issue. If you look in sentry there is (or was) a lot of errors relating to state being updated on unmounted components. I'm not 100% sure, but I think this could also lead to state updates from componentV1 applied to componentV2. Having longer debounce values exposes the issue.

Apart from that your current implementation is flawed in 3 ways:
1. You can't store functions inside state like this. This looks to React that you want to use debounced function as a state setter. Whatever that function returns *will become state* instead of the function itself.
2. If you fix that issue and store the function akin to how refs work, e.g. put `{ current: fn }` in state, it will cause an infinite update loop unless the func argument is wrapped in useCallback, not a particularly user friendly API.
3. The hook doesnt return a callback, it returns undefined and only sets a callback after mount. This is not an expected behavior: useCallback returns a function immediately.

Try this playground: https://codepen.io/i-wont-give-my-name-to-a-machine/pen/mdraJRZ?editors=0010

Reply:
First big point to mention: **This should be the very first hook to be tested in this repo.** It's not trivial and likely to break if we need to change it in future. Others may not understand the issues it solves and refactor it, ...

To point 1: Good point - easy to fix.
To point 2: I don't think there is a way around using `useCallback` for the `fn` argument and I think we already use `useCallback` everywhere together with `useDebouncedCallback`. We need to know when to clean up debounce instances and I think it's most straight forward to use `fn` for that. (Offtopic: When passing callbacks or other data to hooks in libraries, it is quite risky not to memo them at all because the implementation might change and they might be used as deps. I think this is an overall issue with memo in react hooks. We should not use memo everywhere but you sometimes doesn't know what a component or hook you are using expects.)
To point 3: Yeah good point (and easy to fix). Wouldn't want to check for null everytime I use a debounced callback.

I am quite confused by your implementation. You return a new debounced instance on **every** render. That means that you cannot use it as a dep (unexpected imo). I don't think it will catch all calls like expected. Try it yourself in your example:
- Click 'cached handler'
- Increate the counter 
- Click 'cached handler'
- Only the last click will be executed because you cancel out the first call because the component had to render.

I still think we should keep the debounced instance in state. We initialize it in `useState` so it's already there on first render, then we listen for changes on `func` and `delay` and create new instances and cancel out old ones. Sounds like the most simple solution to me.

Reply:
> You return a new debounced instance on every render.
Yep good catch, need to wrap in useCallback for stability. Doesn't break anything but does cause unnecessary rerenders.

> I still think we should keep the debounced instance in state.
I don't think we need state at all. State is needed to cause rerenders when it changes. We already get that from changing fn (external factors) and only need to produce debounced function.

> Only the last click will be executed because you cancel out the first call because the component had to render.
This is probably OK since props may have changed during render in such a way that its no longer displaying the same content at all. Up for debate on this, but afaik that's how every other wrapped debounce worked so far.

Reply:
> This is probably OK since props may have changed during render in such a way that its no longer displaying the same content at all. Up for debate on this, but afaik that's how every other wrapped debounce worked so far.

Think of a component that shows the current time and re-renders each second. Wouldn't make sense there. Let's do a call about this whole topic

Reply:
I imagined it to work more or less like max's code
```
const useDebouncedCallback = (fn, delay = 1000) => {
  // I think it's ok to call debounce on each render
  const wrapped = useCallback(debounce(fn, delay), [fn, delay]);

  // when wrapped function is changed cancel previous one
  // no need in cache ref, useEffect callback already has reference to old function
  useEffect(() => () => wrapped.cancel(), [wrapped]);

  return wrapped;
};
```

Reply:
> re-renders each second
due to how hooks work react might decide to just re-render a component sporadially. that means the debounce time will not be guaranteed. e.g. rerender component every 100ms, while debounce set to 2000ms? fn gets called on unmount via .flush every 100ms

Reply:
```javascript
const useDebouncedCallback = (fn, delay = 1000, deps = []) => {
  const callback = useRef(null);
  callback.current = fn;
  
  const wrapped = useMemo(() => debounce((...args) => {
    callback.current?.(...args);
  }, delay), []);

  useEffect(() => {
    return () => { wrapped.cancel(); };
  }, deps);

  return wrapped;
};
```